### PR TITLE
Second try..  Add check box to enable analysis by user..

### PR DESCRIPTION
**Pull request by matiasdelellis**
Ready to first review, but need help.. 

The only difficult case is when the user deactivates the analysis while it is running.. How do we handle it? 'If ($userEnabled)? everywhere?

### CODE REVIEW

File: js/personal.js [hunk 1]
```diff
@@ -94,19 +106,42 @@ View.prototype = {
...
+            url: OC.generateUrl('apps/facerecognition/setuservalue'),
+            data: {
+                'type': 'enabled',
+                'value': enabled
+            },
+            success: function () {
+                if (enabled) {
+                    OC.Notification.showTemporary(t('facerecognition', 'The analysis is enabled, please be patient, you will soon see your friends here.'));
+                } else {
+                    OC.Notification.showTemporary(t('facerecognition', 'The analysis is disabled, we eliminate all information for the recognition of your friends.'));
```
Comment (stalker314314) [inline js/personal.js]: eliminate -> eliminated? I would even go with "removed all face data"

**Comment by stalker314314** [comment]
> The only difficult case is when the user deactivates the analysis while it is running.. How do we handle it? 'If ($userEnabled)? everywhere?

so, if we are in the middle of the task, who know what can happen, probably some crash due to foreign keys:) We can either delay deletion after task is done (you don't want to do this, if I understand you, you want *immediate* action) or we can add another task, after all other tasks are done, that will check, for all users, if user is disabled and do reset again (just in case there is some leftovers).

File: js/personal.js (hunk 1 above)
Comment (matiasdelellis) [reply js/personal.js]: I think you already suggested it in another branch.. :sweat_smile: 
Ok.

**Comment by matiasdelellis** [comment]
> so, if we are in the middle of the task, who know what can happen, probably some crash due to foreign keys:) 

Exactly .. I have no idea .. but it can not be good .. :see_no_evil: 

> We can either delay deletion after task is done (you don't want to do this, if I understand you, you want immediate action) or we can add another task, after all other tasks are done, that will check, for all users, if user is disabled and do reset again (just in case there is some leftovers).

I guess for a first version, we can delete the scheduled task ..
And in the notification of the fronted, say something like this:
 * _They will be completely eliminated in the next analysis, after that, if you re-enable it, you will start from scratch._

File: lib/Controller/SettingController.php [hunk 2]
```diff
@@ -69,6 +70,8 @@ public function setUserValue($type, $value) {
 			case 'enabled':
 				if ($value === 'false')
 					$this->faceManagementService->resetAllForUser($this->userId);
+				else
```
Comment (stalker314314) [inline lib/Controller/SettingController.php]: nice catch!:)

**Comment by matiasdelellis** [comment]
Hi @stalker314314 
I still have little time, but at least I will try to finish this. 

About the backend, I think it's all necessary and it's fine.. but please review it again.

**Comment by matiasdelellis** [comment]
Well,
Still is necessary to correct the tests, since as default is disabled, it never works ...

File: js/personal.js [hunk 3]
```diff
@@ -94,19 +106,42 @@ View.prototype = {
...
+            url: OC.generateUrl('apps/facerecognition/setuservalue'),
+            data: {
+                'type': 'enabled',
+                'value': enabled
+            },
+            success: function () {
+                if (enabled) {
+                    OC.Notification.showTemporary(t('facerecognition', 'The analysis is enabled, please be patient, you will soon see your friends here.'));
+                } else {
+                    OC.Notification.showTemporary(t('facerecognition', 'The analysis is disabled. In the next execution your data will be completely eliminated.'));
```
Comment (stalker314314) [inline js/personal.js]: consider rewording `your data will be completely eliminated` to something less scary, for example `The analysis is disabled. We will remove data about all your recognized faces.` (this way, it seems like some user data might be removed, and "eliminated" seems like something that Daleks from Dr. Who would say:D)

Comment (stalker314314) [reply js/personal.js]: or, as you said in other place, "You will lose all the information analyzed"

File: lib/Controller/PersonController.php [hunk 4]
```diff
@@ -63,6 +63,8 @@ public function index() {
 		$model = intval($this->config->getAppValue('facerecognition', 'model', AddDefaultFaceModel::DEFAULT_FACE_MODEL_ID));
 
 		$resp = array();
+		$resp['enabled'] = $this->config->getUserValue($this->userId, 'facerecognition', 'enabled', false);
```
Comment (stalker314314) [inline lib/Controller/PersonController.php]: if not enabled, should we send persons to client at all? Why just not "fake" it and send empty array?

File: lib/BackgroundJob/Tasks/DisabledUserRemovalTask.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+use OCP\IUser;
+
+use OCA\FaceRecognition\BackgroundJob\FaceRecognitionBackgroundTask;
+use OCA\FaceRecognition\BackgroundJob\FaceRecognitionContext;
+use OCA\FaceRecognition\Db\ImageMapper;
+use OCA\FaceRecognition\Migration\AddDefaultFaceModel;
+
+use OCA\FaceRecognition\FaceManagementService;
+
+/**
```
Comment (stalker314314) [inline lib/BackgroundJob/Tasks/DisabledUserRemovalTask.php]: is comment stale?

File: lib/Controller/PersonController.php (hunk 4 above)
Comment (matiasdelellis) [reply lib/Controller/PersonController.php]: Oh .. You're right... :+1:

File: lib/BackgroundJob/Tasks/DisabledUserRemovalTask.php (hunk 5 above)
Comment (matiasdelellis) [reply lib/BackgroundJob/Tasks/DisabledUserRemovalTask.php]: Ohh.. Thanks. :disappointed:

**Comment by matiasdelellis** [comment]
Please,
Last review and merge it.. :smiley: 

With the latest commits, improve the user experience.. . and we can continue with other things.

File: lib/BackgroundJob/Tasks/DisabledUserRemovalTask.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+				$eligable_users[] = $user->getUID();
+			});
+		} else {
+			$eligable_users[] = $this->context->user->getUID();
+		}
+
+		// Reset user datas if needed.
+		foreach($eligable_users as $userId) {
+			$userEnabled = $this->config->getUserValue($userId, 'facerecognition', 'enabled', 'false');
+			$imageCount = $this->imageMapper->countUserImages($userId, $model);
```
Comment (stalker314314) [inline lib/BackgroundJob/Tasks/DisabledUserRemovalTask.php]: Ideally, you want to check here for personCount too (persons are top-level entity and can exist without any image or face). I mean, normally, if they exist and there is no images, it's a bug, but...if there is a bug, no way user can delete those persons.

Comment (matiasdelellis) [reply lib/BackgroundJob/Tasks/DisabledUserRemovalTask.php]: Let's reproduce the error before adding ugly unnecessary fixes.

**Pull request merged**